x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/EsField.java

```diff
@@ -77,18 +90,27 @@ private DataType readDataType(StreamInput in) throws IOException {
         return DataType.readFrom(name);
     }
 
+    public static <A extends EsField> A readFrom(StreamInput in) throws IOException {
+        return ((PlanStreamInput) in).readEsFieldWithCache();
+    }
+
     @Override
     public void writeTo(StreamOutput out) throws IOException {
+        if (((PlanStreamOutput) out).writeEsFieldCacheHeader(this)) {
+            writeContent(out);
+        }
+    }
+
+    protected void writeContent(StreamOutput out) throws IOException {
```

Comment:
I think it should have a javadoc saying it's overridden by subclasses for specific serialization.